[mupdf] Add custom allocator to avoid having fuzzer kill process (#1830)

sebras opened the pull request:
This fixes oss-fuzz #5679 and oss-fuzz #7803 for the mupdf project.

I have tested this locally and works fine; detecting 
allocations that in total are above 1Gbyte.

Dor1s commented on projects/mupdf/pdf_fuzzer.cc:
```diff
@@ -51,5 +149,7 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
 
   fz_drop_context(ctx);
 
+  fprintf(stderr, "total allocation at end of execution: %" PRIu64 " bytes\n", total);
```
It's recommended not to print anything from a fuzz target, especially inside LLVMFuzzerTestOneInput function which is executed a huge number of times.

oliverchang commented on projects/mupdf/pdf_fuzzer.cc:
```diff
@@ -17,11 +17,101 @@
+	if (size > SIZE_MAX - ALIGNMENT)
+		return NULL;
+
+	if (total + size + ALIGNMENT > MAX_ALLOCATION)
```
this addition can potentially overflow and cause the check to incorrectly fail. should this be properly handled here?

oliverchang commented on projects/mupdf/pdf_fuzzer.cc:
```diff
@@ -17,11 +17,101 @@
+	old = ((char *) old) - ALIGNMENT;
+	memcpy(&oldsize, old, sizeof(oldsize));
+
+	if (total - oldsize + size > MAX_ALLOCATION)
```
same here

sebras commented:
Good point. I have fixed those potential sources of overflow I believe. Additionally the code was being ambivalent about whether to include `ALIGNMENT` in `total` or not (it did in comparisons but not when increasing/decreasing `total`). I fixed that too. I re-tested on the reproducers from oss-fuzz #5679 and oss-fuzz #7803 locally and the allocator returns `NULL` before the fuzzer aborts the test, just like intended.

oliverchang commented:
Great, thanks! Let's give this a shot.
